sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/ReleaseDownloadListener.java

```diff
@@ -0,0 +1,179 @@
+        SharedPreferencesManager.putLong(PREFERENCE_KEY_DOWNLOAD_TIME, enqueueTime);
+    }
+
+    @Override
+    public boolean onProgress(final long currentSize, final long totalSize) {
+        AppCenterLog.verbose(LOG_TAG, "downloadedBytes=" + currentSize + " totalBytes=" + totalSize);
+
+        /* If file size is known update downloadProgress bar. */
+        if (mProgressDialog != null && totalSize >= 0) {
+            HandlerUtils.runOnUiThread(new Runnable() {
+
+                @Override
+                public void run() {
+
+                    /* When we switch from indeterminate to determinate */
+                    if (mProgressDialog.isIndeterminate()) {
+
+                        /* Configure the progress dialog determinate style. */
+                        mProgressDialog.setProgressPercentFormat(NumberFormat.getPercentInstance());
+                        mProgressDialog.setProgressNumberFormat(mContext.getString(R.string.appcenter_distribute_download_progress_number_format));
+                        mProgressDialog.setIndeterminate(false);
+                        mProgressDialog.setMax((int) (totalSize / MEBIBYTE_IN_BYTES));
+                    }
+                    mProgressDialog.setProgress((int) (currentSize / MEBIBYTE_IN_BYTES));
+                }
+            });
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean onComplete(@NonNull Uri localUri) {
+        Intent intent = getInstallIntent(localUri);
+        if (intent.resolveActivity(mContext.getPackageManager()) == null) {
+            return false;
+        }
+
+        /* Check if app should install now. */
+        if (!Distribute.getInstance().notifyDownload(mReleaseDetails, intent)) {
+
+            /*
+             * This start call triggers strict mode in U.I. thread so it
+             * needs to be done here without synchronizing
+             * (not to block methods waiting on synchronized on U.I. thread)
+             * so yes we could launch install and SDK being disabled...
+             *
+             * This corner case cannot be avoided without triggering
+             * strict mode exception.
+             */
+            AppCenterLog.info(LOG_TAG, "Show install UI now intentUri=" + intent.getData());
+            mContext.startActivity(intent);
+            if (mReleaseDetails.isMandatoryUpdate()) {
+                Distribute.getInstance().setInstalling(mReleaseDetails);
+            }
+            storeReleaseDetails(mReleaseDetails);
+        }
+        return true;
+    }
+
+    @Override
+    public void onError(@NonNull String errorMessage) {
+        AppCenterLog.error(LOG_TAG, errorMessage);
+        Distribute.getInstance().completeWorkflow(mReleaseDetails);
+    }
+
+    /**
+     * Show download progress. Only used for mandatory updates.
+     */
+    @SuppressWarnings({"deprecation", "RedundantSuppression"})
+    ProgressDialog showDownloadProgress(Activity foregroundActivity) {
```

Comment:
What happened to the previous check which was testing activity was in foreground?
I'm not following all the changes because the PR is so big.